File: apps/bhub/lib/services/bhub_service.dart
import 'package:dio/dio.dart';

import '../models/bhub_model.dart';

/// Service for handling BHub API operations
class BHubService {
  final Dio _dio = Dio(BaseOptions(
    baseUrl: 'http://localhost:8080',
    connectTimeout: const Duration(seconds: 5),
    receiveTimeout: const Duration(seconds: 3),
  ));

  /// Get a list of available BHubs
  Future<List<BHubResponse>> listBHubs() async {  // Removed location parameter
    try {
      final response = await _dio.get('/bhub/list');  // Changed to match BE endpoint
      
      if (response.statusCode == 200) {
        final List<dynamic> data = response.data;
        return data.map((json) => BHubResponse.fromJson(json as Map<String, dynamic>)).toList();
      } else {
        throw Exception('Failed to load BHubs: ${response.statusCode}');
      }
    } catch (e) {
      // For demo purposes, return mock data
      return _getMockBHubs();
    }
  }

  /// Get a specific BHub by ID
  Future<BHubResponse> getBHub(String bhubId) async {
    try {
      final response = await _dio.get('/bhub/$bhubId');
      
      if (response.statusCode == 200) {
        return BHubResponse.fromJson(response.data as Map<String, dynamic>);
      } else {
        throw Exception('Failed to load BHub: ${response.statusCode}');
      }
    } catch (e) {
      // For demo purposes, return mock data
      return _getMockBHubs()[0];
    }
  }

  /// Join a BHub
  Future<BHubJoinResponse> joinBHub(String bhubId, String userId) async {
    try {
      final request = BHubJoinRequest(
        bhubId: bhubId,
        userId: userId,
      );
      
      final response = await _dio.post(
        '/bhub/$bhubId/join',
        data: request.toJson(),
      );
      
      if (response.statusCode == 200) {
        return BHubJoinResponse.fromJson(response.data as Map<String, dynamic>);
      } else {
        throw Exception('Failed to join BHub: ${response.statusCode}');
      }
    } catch (e) {
      // For demo purposes, return mock response
      return BHubJoinResponse(
        message: 'Successfully joined BHub',
        totalPrice: 15.0,
      );
    }
  }

  /// Create a new BHub
  Future<BHubCreateResponse> createBHub(BHubCreateRequest request) async {
    try {
      final response = await _dio.post(
        '/bhub/create',
        data: request.toJson(),
      );
      
      if (response.statusCode == 201) {
        return BHubCreateResponse.fromJson(response.data as Map<String, dynamic>);
      } else {
        throw Exception('Failed to create BHub: ${response.statusCode}');
      }
    } catch (e) {
      // For demo purposes, return mock response
      return BHubCreateResponse(
        bhubId: 'mock-bhub-id',
        pricePerPerson: request.priceTotal / request.maxMembers,
      );
    }
  }

  // Mock data for demo purposes
  List<BHubResponse> _getMockBHubs() {
    final now = DateTime.now();
    return [
      BHubResponse(
        id: 'bhub-1',
        location: 'New York',
        timeSlot: DateTime(now.year, now.month, now.day + 1, 18, 0),
        currentMembers: 3,
        maxMembers: 6,
        pricePerPerson: 15.0,
        host: 'John Doe',
        status: 'open',
      ),
      BHubResponse(
        id: 'bhub-2',
        location: 'Los Angeles',
        timeSlot: DateTime(now.year, now.month, now.day + 2, 19, 30),
        currentMembers: 4,
        maxMembers: 4,
        pricePerPerson: 18.0,
        host: 'Jane Smith',
        status: 'full',
      ),
      BHubResponse(
        id: 'bhub-3',
        location: 'Chicago',
        timeSlot: DateTime(now.year, now.month, now.day + 3, 17, 0),
        currentMembers: 2,
        maxMembers: 8,
        pricePerPerson: 12.5,
        host: 'Mike Johnson',
        status: 'open',
      ),
      BHubResponse(
        id: 'bhub-4',
        location: 'Houston',
        timeSlot: DateTime(now.year, now.month, now.day + 1, 20, 0),
        currentMembers: 1,
        maxMembers: 6,
        pricePerPerson: 14.0,
        host: 'Sarah Williams',
        status: 'open',
      ),
    ];
  }
}

File: apps/api/services/bhub_service.go
package services

import (
	"context"
	"errors"
	"time"

	"github.com/ducnmm/bhub/apps/api/models"
	"github.com/ducnmm/bhub/apps/api/repositories"
)

// BHubService handles BHub operations
type BHubService struct {
	BHubRepo *repositories.BHubRepository
	UserRepo *repositories.UserRepository
}

// NewBHubService creates a new BHub service instance
func NewBHubService(bhubRepo *repositories.BHubRepository, userRepo *repositories.UserRepository) *BHubService {
	return &BHubService{
		BHubRepo: bhubRepo,
		UserRepo: userRepo,
	}
}

// CreateBHub creates a new BHub
func (s *BHubService) CreateBHub(ctx context.Context, req *models.BHubCreateRequest) (*models.BHubCreateResponse, error) {
	// Verify host exists
	host, err := s.UserRepo.GetByID(ctx, req.HostID)
	if err != nil || host == nil {
		return nil, errors.New("invalid host ID")
	}

	// Create BHub
	bhub := &models.BHub{
		HostID:     req.HostID,
		Location:   req.Location,
		TimeSlot:   req.TimeSlot,
		MinMembers: req.MinMembers,
		MaxMembers: req.MaxMembers,
		PriceTotal: req.PriceTotal,
	}

	// Save BHub
	if err := s.BHubRepo.Create(ctx, bhub); err != nil {
		return nil, err
	}

	return &models.BHubCreateResponse{
		BHubID:         bhub.ID,
		PricePerPerson: bhub.PricePerPerson,
	}, nil
}

// JoinBHub handles a user joining a BHub
func (s *BHubService) JoinBHub(ctx context.Context, req *models.BHubJoinRequest) (*models.BHubJoinResponse, error) {
	// Get BHub
	bhub, err := s.BHubRepo.GetByID(ctx, req.BHubID)
	if err != nil {
		return nil, err
	}

	// Verify user exists
	user, err := s.UserRepo.GetByID(ctx, req.UserID)
	if err != nil || user == nil {
		return nil, errors.New("invalid user ID")
	}

	// Check if BHub is full
	if len(bhub.Members) >= bhub.MaxMembers {
		return nil, errors.New("BHub is full")
	}

	// Check if BHub is still open
	if bhub.Status != "open" {
		return nil, errors.New("BHub is not open for joining")
	}

	// Add member to BHub
	if err := s.BHubRepo.AddMember(ctx, req.BHubID, req.UserID); err != nil {
		return nil, err
	}

	return &models.BHubJoinResponse{
		Message:    "Successfully joined BHub",
		TotalPrice: bhub.PricePerPerson,
	}, nil
}

// ListBHubs retrieves available BHubs
func (s *BHubService) ListBHubs(ctx context.Context) ([]*models.BHubResponse, error) {  // Removed location parameter
    // Get BHubs from current time onwards
    bhubs, err := s.BHubRepo.List(ctx, time.Now())
    if err != nil {
        return nil, err
    }

    var responses []*models.BHubResponse
    for _, bhub := range bhubs {
        // Get host information
        host, err := s.UserRepo.GetByID(ctx, bhub.HostID)
        if err != nil {
            continue
        }

        responses = append(responses, &models.BHubResponse{
            ID:             bhub.ID,
            Location:       bhub.Location,
            TimeSlot:       bhub.TimeSlot,
            CurrentMembers: len(bhub.Members),
            MaxMembers:     bhub.MaxMembers,
            PricePerPerson: bhub.PricePerPerson,
            Host:           host.Name,
            Status:         bhub.Status,
        })
    }

    return responses, nil
}

// GetBHub retrieves a specific BHub by ID
func (s *BHubService) GetBHub(ctx context.Context, id string) (*models.BHubResponse, error) {
	// Get BHub
	bhub, err := s.BHubRepo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}

	// Get host information
	host, err := s.UserRepo.GetByID(ctx, bhub.HostID)
	if err != nil {
		return nil, err
	}

	return &models.BHubResponse{
		ID:             bhub.ID,
		Location:       bhub.Location,
		TimeSlot:       bhub.TimeSlot,
		CurrentMembers: len(bhub.Members),
		MaxMembers:     bhub.MaxMembers,
		PricePerPerson: bhub.PricePerPerson,
		Host:           host.Name,
		Status:         bhub.Status,
	}, nil
}
